Assignment Endpoint & other fixes

simpslandyy opened the pull request:
@freeatnet @michellemtchai please review

## New
Migration - added a column to the assignment table called **hours**
Methods - implemented methods for assignment endpoint
Test Cases - some test cases for the assignment endpoint, pending test cases for application endpoint 

_________________________________________________________________________________
## Methods
#### index #GET   
     /assignments - returns a list of applicant including their assignment data
     /applicants/:applicant_id/assignments - returns an assignments given an applicant ID

#### show #GET     
    /assignments/:id - returns an assignment given an assignment ID
    /applicants/:applicant_id/assignments/:id - returns an assignment given an ID, 
    for a given applicant ID

#### create #POST  
     creates an assignment between an applicant and a position.
     returns status 209 if tries to create a duplicate

     /applicants/:applicant_id/assignments
     body { round_id,  hours,  position_id }

#### update #PATCH 
    updates the hours column for an applicant given assignment ID and applicant ID

    /applicants/:applicant_id/assignment/:id
    body { hours }

#### destroy #DELETE
    removes (unassigns) an assignment record give an applicant ID and assignment ID
    /applicants/:applicant_id/assignment/:id

michellemtchai commented:
- The assignment test gave a bit of an issue because the `db:migrate` was only done on the development side and the test wasn't. So, in order to run the test, I had to run `docker-compose run -eRAILS_ENV=test rails-app rake db:migrate`. 
- The following is the result from the test:
````
AssignmentsController
  GET #index
    when not passed an applicant ID
      lists assignments
    when passed an applicant ID
      lists assignments given an integer applicant ID
    lists assignments given a non-integer applicant ID
      returns 404
  POST #create
    example at ./spec/controllers/assignments_controller_spec.rb:171 (PENDING: Not yet implemented)
  PATCH #update
    example at ./spec/controllers/assignments_controller_spec.rb:174 (PENDING: Not yet implemented)
  DELETE #destroy
    example at ./spec/controllers/assignments_controller_spec.rb:178 (PENDING: Not yet implemented)
  GET #show
    when passed a non-integer ID
      returns 404
      returns 404, given an applicant ID
    when passed an integer ID
      returns the assignment
      returns the assignment for the given applicant_id

Pending: (Failures listed here are expected and do not affect your suite's status)

  1) AssignmentsController POST #create 
     # Not yet implemented
     # ./spec/controllers/assignments_controller_spec.rb:171

  2) AssignmentsController PATCH #update 
     # Not yet implemented
     # ./spec/controllers/assignments_controller_spec.rb:174

  3) AssignmentsController DELETE #destroy 
     # Not yet implemented
     # ./spec/controllers/assignments_controller_spec.rb:178


Finished in 0.24889 seconds (files took 1.5 seconds to load)
10 examples, 0 failures, 3 pending
````
Are these the expected result?

michellemtchai commented:
- `GET /assignment` gave an array consisting of the following attributes:
  - id": 1,
  - utorid
  - student_number
  - first_name
  - last_name
  - email
  - phone
  - address
  - created_at
  - updated_at
  - assignments: this attribute appeared to be an array


- `GET /applicants/:applicant_id/assignments` returned an object with the same attributes as listed above


- `POST /applicants/:applicant_id/assignments` returned the same object format as `GET 
  /applicants/:applicant_id/assignments` and the attributes of assignments are listed as:
  - id
  - applicant_id
  - position_id
  - round_id
  - created_at
  - updated_at
  - hours

The `round_id` in this post option bothered me because I simply entered a random number that was not an existing `round_id` in the Courses model. So, you might want to throw an error if the `round_id` is blatantly made up. Otherwise, the posting of identical assignment values does return a `status: 209`. Posting an incomplete set of body attributes (without either `hours` or `round_id`) does not seems to raise any error. This might be something you want to address as a `status: 404` because I was able to insert an Assignment with no `round_id` and another one without `hours`. Posting non-existent `position_id` and not giving a `position_id` both gives a Rails error with `ActiveRecord::RecordInvalid: Validation failed: Position must exist`. You might want to catch this issue.

- `PATCH /applicants/:applicant_id/assignment/:id` does update the `hours` attributes correctly. Extra attributes added to the PATCH doesn't seem to cause any issues either. The patching of non-existent assignment with `assignment_id` returns a `status: 404`

- `GET /applicants/:applicant_id/assignments/:id` list all attributes of an Assignment with `id` and when the `id` is invalid it returns `status: 404`

- `DELETE /applicants/:applicant_id/assignment/:id` deletes the desired Assignment by its `id` and when given an invalid `id` it returns a`status: 404`.

Overall, the routes works fine. There was a bit of issue with migrating the test db and some exceptions you may want to catch for `POST /applicants/:applicant_id/assignments`.

freeatnet commented on db/migrate/20170609201342_add_hours_to_assignments.rb:
```diff
@@ -0,0 +1,5 @@
+class AddHoursToAssignments < ActiveRecord::Migration[5.1]
+  def change
+    add_column :assignments, :hours, :integer
```
💯

freeatnet commented on spec/controllers/applications_controller_spec.rb:
```diff
@@ -116,6 +116,10 @@
 
+      # it "returns the application" do
+      #   pending "in the works"
+      # end
```
😆

freeatnet commented on app/controllers/assignments_controller.rb:
```diff
@@ -0,0 +1,93 @@
+      round_id, hours, position_id
+
+      creates an assignment between an applicant and a position.
+      returns status 209 if tries to create a duplicate
```
Could you explain your reasoning of using a success-range code, rather than something like `422 Unprocessable Entity` or `409 Conflict`?

simpslandyy replied:
oh this was my bad, I meant to say 409 conflict.

freeatnet commented on app/controllers/assignments_controller.rb:
```diff
@@ -0,0 +1,93 @@
+
+    unless has_assignment
+      @applicant = Applicant.find(params[:applicant_id])
+      @applicant.assignments.create!(assignment_params)
```
We should probably report errors encountered here. The common pattern is
```
assignment = @applicant.assignment.build(assignment_params)
if assignment.save
  render json: assignment.to_json
else
  render json: assignment.errors.to_hash(true), status: :unprocessable_entity
end
```

freeatnet commented on app/controllers/assignments_controller.rb:
```diff
@@ -0,0 +1,93 @@
+    @applicant = Applicant.find(params[:applicant_id])
+    @assignment = @applicant.assignments.find(params[:id])
+
+    @assignment.update!(hours: params[:hours])
```
Ditto.

freeatnet commented on app/controllers/assignments_controller.rb:
```diff
@@ -0,0 +1,93 @@
+
+  private
+    def assignment_params
+      params.permit(:position_id, :round_id, :hours)
```
Are we leaving figuring out of `round_id` to the application then?

freeatnet commented on spec/controllers/assignments_controller_spec.rb:
```diff
@@ -0,0 +1,181 @@
+      end
+    end
+
+    context "lists assignments given a non-integer applicant ID" do
```
"Non-integer", but more importantly, "invalid" or "nonexistent" :)

freeatnet commented on spec/controllers/assignments_controller_spec.rb:
```diff
@@ -0,0 +1,181 @@
+        )
+    end
+
+    context "when passed an integer ID" do
```
"the assignment ID"?

freeatnet commented:
A rebase is also needed for this branch.

simpslandyy commented:
@freeatnet, @michellemtchai  - okay so I've fixed up the assignment and application endpoints, and also their test cases.
I could use a bit of feedback for the test cases though, so if you don't mind just looking over them.
The POST request for creating an assignment when a POSITION is CLOSED, is failing. I'm almost 100% sure it's my logic, I'm just not sure why.

Also, is there a way to type check parameters?
2/3 of the assignment test cases are failing because when hours is a non-integer, it still creates/updates the record the database (hours just gets set to 0).

Thank you :)

michellemtchai commented:
Out of all the tests, the only ones that failed was the `assignments_controller_spec` and the `positions_controller_spec`. The following was the error thrown  for `assignments_controller_spec`:
```
AssignmentsController
Failures:

  1) AssignmentsController PATCH #update hours is a non integer parameter
     Failure/Error: expect(response.status).to eq(404)
     
       expected: 404
            got: 200
     
       (compared using ==)
     # ./spec/controllers/assignments_controller_spec.rb:305:in `block (3 levels) in <top (required)>'

  2) AssignmentsController POST #create when position is open & passed a number of hours, given a non-integer hours parameter
     Failure/Error: expect(response.status).to eq(404)
     
       expected: 404
            got: 200
     
       (compared using ==)
     # ./spec/controllers/assignments_controller_spec.rb:235:in `block (5 levels) in <top (required)>'

  3) AssignmentsController POST #create when position is closed throws a 404 status
     Failure/Error: expect(response.status).to eq(404)
     
       expected: 404
            got: 200
     
       (compared using ==)
     # ./spec/controllers/assignments_controller_spec.rb:243:in `block (4 levels) in <top (required)>'

18 examples, 3 failures
```
The `positions_controller_spec` failed because it wasn't up-to-date with `23-remove-course-model` branch. You will need to `rebase` to get this part working.

simpslandyy commented:
@michellemtchai - Are the test cases for positions something you modified yesterday after I rebased, or before I rebased? & what are the failed test cases? @freeatnet I'm a bit confused with why I need to constantly be rebasing with a branch in-progress, isn't that really risky in comparisons to rebasing with something stable.

freeatnet commented:
@simpslandyy From my perspective, we attempted to get a little bit ahead of the changes that are gonna come in via #43. You don't have to keep rebasing — hopefully, #43 drops in master soon enough.

simpslandyy commented:
@michellemtchai @freeatnet please review again . I think this is it from me for the assignment endpoint.
All tests should pass.

michellemtchai commented:
@simpslandyy I had to run `docker-compose run -eRAILS_ENV=development rails-app rake db:migrate` to get the routes working, again. Otherwise, all the tests are passing and all routes works according to your specifications.

chrislessard commented:
Looks good to me
